src/practitest_firecracker/api.clj

```diff
@@ -0,0 +1,235 @@
+(ns practitest-firecracker.api
+  (:require
+    [clj-http.client :as http]
+    [clojure.string :as string]
+    [throttler.core :refer [fn-throttler]]
+    [clojure.tools.logging :as log]
+    [practitest-firecracker.const :refer :all]
+    [practitest-firecracker.utils :refer [exit group-errors pformat]]))
+
+(def backoff-timeout "Backoff timeout in seconds" 20)
+(def max-attempts "Number of attempts to run" 10)
+(def timeout-between-attempts "Timeout in seconds between attempts" 1)
+
+(defn create-api-throttler [rate]
+  (let [fn-th (fn-throttler rate :minute)]
+    fn-th))
+
+(defn build-uri [base-uri resource-uri-template & params]
+  (apply format (str base-uri resource-uri-template) params))
+
+(defn throw-api-exception [ex-info status body uri]
+  (exit status (group-errors body)))
+
+(defn api-call [{:keys [credentials uri method query-params form-params]}]
+  (assert (not (and query-params form-params))
+          "both `query-params` and `form-params` can't be specified")
+  (loop [results []
+         uri uri
+         attempts max-attempts
+         params (cond-> {:basic-auth       credentials
+                         :throw-exceptions false
+                         :as               :json}
+                        query-params (assoc :query-params (conj query-params {:source "firecracker" :firecracker-version fc-version}))
+                        form-params (assoc :form-params (conj form-params {:source "firecracker" :firecracker-version fc-version}) :content-type :json))]
+    (if (nil? uri)
+      results
+      (let [{:keys [status body]} (method uri params)]
+        (if (> attempts 0)
+          (case status
+            504 (do
+                  ;; load balancer freaking out, lets try again
+                  (Thread/sleep (* timeout-between-attempts 1000))
+                  (recur results uri (dec attempts) params))
+            502 (do
```

Comment:
502, 500, 503 can be handled with single branch, something like

```
(500 502 503) (do 
              (log/warnf "%s responded with %s - %s more attempts" uri status attempts)
                  (Thread/sleep (* timeout-between-attempts 1000))
                  (recur results uri (dec attempts) params))
```

Reply:
thank you, changed